File: 0015-3sum/solution.py
class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        if len(nums) < 3:
            return []
        solutions = []
        nums.sort()
        for i in range(len(nums)):
            value = []
            if i > 0 and nums[i - 1] == nums[i]:
                continue
            l, r = i + 1, len(nums) - 1
            while l < r:
                if nums[i] + nums[r] + nums[l] > 0:
                    r -= 1
                elif nums[i] + nums[r] + nums[l] < 0:
                    l += 1
                else:
                    solutions.append([nums[i], nums[l], nums[r]])
                    l+=1
                    while nums[l] == nums[l - 1] and l < r:
                        l+=1
        return solutions

File: 0015-3sum/solution.go
import "slices"
func threeSum(nums []int) [][]int {
    ans := [][]int{}
    slices.Sort(nums)
    for i := 0; i < len(nums) - 2; i++ {
        if i > 0 && nums[i] == nums[i - 1] {
            continue
        }
        for left, right := i + 1, len(nums) - 1; left < right; {
            sum := nums[i] + nums[left] + nums[right]
            if sum < 0 {
                left++
            } else if sum > 0 {
                right--
            } else {
                ans = append(ans, []int{nums[i], nums[left], nums[right]})
                prevLeft := nums[left]
                prevRight := nums[right]
                left++
                right--
                for left < right && nums[left] == prevLeft {
                    left++
                }
                for left < right && nums[right] == prevRight {
                    right--
                }
            }
        }
    }
    return ans
}
